### ISSUE TITLE
Make HikariCP loom friendly

### ISSUE THREAD

**Issue by cogman** [brettwooldridge/HikariCP#1463]
Currently, HikariCP uses the synchronized keyword in a few places. Unfortunately, loom may (or may not) support the synchronized keyword pausing a fiber, instead it may end up pausing the fiber carrier thread (it is an open question).

See: https://cr.openjdk.java.net/~rpressler/loom/Loom-Proposal.html

And where those happen: https://github.com/brettwooldridge/HikariCP/search?q=synchronized&unscoped_q=synchronized

Fortunately there is an easy solution here, instead of using the synchronized keyword, use j.u.c locks such as https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html

These APIs will be guaranteed to pause fibers when they are hit.

**Comment by Glusk** [comment]
<del>IIUC, there will be no need for a thread pool anymore.<br>
[In one of the podcasts](https://youtu.be/EDVesJ-yJ6U?list=PLX8CzqL3ArzV_hXbRevwzrXSMcGNzhxiZ&t=1137) introducing the technology, Ron Pressler suggests using a combination of a semaphore and virtual threads instead of a connection pool.</del>

---

🤦 I just realised there's a difference between a `Thread` pool and a `Connection` pool. I'm sorry for polluting the issue.

**Comment by cogman** [comment]
@Glusk different problems.

Hikari doesn't manage a thread pool (sort of... not around connections at least).  Instead, Hikari is managing network connections or JDBC connections to the database.  You'll still need libraries like hikari to ensure connections are distributed quickly.

The two expensive parts of connections are the establishment cost and the overhead they introduce onto the server.  Pooling connections puts a hard limit on both.  A semaphore would ensure you don't overwhelm the database server with a bunch of connections but wouldn't solve the expense of connection establishment.

Hikari solves both problems.

What loom solves is if you need a bunch of concurrent requests talking to the DB you don't need to manage the number of threads needed to make those requests.  Loom keeps the OS thread count down which ultimately helps with context switching times and memory requirements.

In other words, you could have a million threads waiting on 20 database connections and that'd be fine.

**Comment by Glusk** [comment]
@cogman 

> A semaphore would ensure you don't overwhelm the database server with a bunch of connections **but wouldn't solve the expense of connection establishment.**

Because opening a new database connection is a bit more than simply opening a new TCP/IP connection to the server. Fair enough. :)

**Comment by cogman** [comment]
@Glusk 

Pretty much, but also remember that TCP/IP connections also aren't cheap.  Absolutely DB connections are more expensive because they usually have some handshake aspect and auth can be somewhat expensive.  But even still, spinning up a bunch of TCP connections can be costly.

There's a reason browsers limit the number of Http1.1 connections to ~6 rather than just doing a new connection per resource request.  The main innovation of Http2 was multiplexing over a single TCP connection.

I think the only place where a more unbounded access of a constrained resource would make sense is file IO.  Particularly with spinning disks, sending a bunch of concurrent file IO requests would allow the OS to better schedule where to put the disk head.

When talking about loom's benefit for an Http server, the main advantage is you cut down the context switching and memory requirements for the server.  So servicing 1 million concurrent requests can ultimately process them based on when resources become available without hogging the system.  You'd still want to limit the number of concurrently processed requests to avoid blowing out the heap, but you could make that whole process be constrained on the network IO rather than the number of threads that can be reasonably spun up.

**Comment by Glusk** [comment]
@cogman 

Aside:
> ***A server can handle upward of _a million_ concurrent open sockets**, yet the operating system cannot efficiently handle more than a few thousand active (non-idle) threads.*
> Source: [State of Loom](http://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part1.html)

**Comment by cogman** [comment]
Yup, that's all correct.

Socket churn can be a bigger issue, however, than the raw number of opened sockets.

This problem has actually struck my company a couple of times.

http://www.serverframework.com/asynchronousevents/2011/01/time-wait-and-its-design-implications-for-protocols-and-scalable-servers.html

Long story short, high TCP connection churn can exhaust the number of available ports in an OS due to the timed wait state.

So while handling a bunch of open sockets isn't really a problem, creating and destroying a bunch of sockets is.

Further, because of TCP slow start, you pay a time penalty every time a new TCP connection is established (The protocol tries to establish how congested the network to attempt to avoid overwhelming it).

https://blog.stackpath.com/tcp-slow-start/

All this is to say, the most efficient use of hardware and network resources is long lived TCP connections that are reused.  There's not necessarily a problem having a million of those connections, but there is a problem if you churn through those connections quickly.  Ideally, the protocol you are using is one that multiplexes over a single TCP connection (such as HTTP2).  That'll give you the highest throughput with the least amount of resources on both the client and server.

Most database connection protocols get this wrong, IMO, primarily because it's simpler for them to just have a bunch of concurrent connections rather than multiplexing requests over a single connection.

**Comment by Blquinn** [comment]
Hello, I wanted to revive this thread now that Java 19 has been release with loom in preview.

There is pretty extensive explanations of the implications of using synchronized blocks from virtual threads in https://openjdk.org/jeps/425.

@cogman's original points about synchronized blocks pinning carrier threads and the use of ReentrantLock to alleviate the issue are still accurate in the current state of loom. However it does mention that if synchronized is just being used for short lived operations on a region of memory then it's probably okay. Where it would be really bad is if you did some longer running, blocking operation in a synchronized block.

**Comment by hyeongguen-song** [comment]
Hi, @brettwooldridge 
Is there any progress on this issue?

**Comment by boris-petrov** [comment]
This is a non-issue since Java 24. In six months Java 25 will be an LTS release and will also contain a fix for this issue. I personally don't think it's sensible to spend time changing the library for something that is already not a problem with the latest Java version.

**Comment by hyeongguen-song** [comment]
@boris-petrov 
JAVA 21 is the most recent LTS version.
24, 25, etc. are not LTS versions.

**Comment by boris-petrov** [comment]
[Java 25 will be an LTS release](https://www.oracle.com/java/technologies/java-se-support-roadmap.html).

**Comment by hyeongguen-song** [comment]
@boris-petrov 
Do you think that projects currently using Java 21 will soon migrate to 25?
I don't know if it's a good idea to not change it because it was fixed in a version that hasn't even been released yet.

**Comment by boris-petrov** [comment]
As I said, it's already fixed in Java 24 which has been released for a week. I'm using it in production (which is totally recommended by Oracle - there is no difference between LTS and non-LTS versions in that sense). Migrating from Java 21 to 25 should be a one-line change for most projects.

I'm not against "fixing" this issue for users of older Java versions. I'm just saying that I would absolutely understand the maintainers if they don't want to. It's a matter of priorities and time.

**Comment by greenhost87** [comment]
I got this problem on Zulu jdk 24... My service stop working right after first API request, somewhere in database call, did anyone have any suggestion???

**Comment by boris-petrov** [comment]
JDK 24 can still deadlock in `static` initializers of classes. There virtual threads pin their carrier threads so you could be hitting something like that. I personally did a few times.

**Comment by domdorn** [comment]
I'm sorry to revive this thread but I couldnt find any documentation for this... How do I have to configure HikariCP so that it uses Looms virtual threads to do JDBC ? Or is HikariCP the wrong place to look at this? I'm running Java 21 latest GraalVM CE.  Thanks!

**Comment by brettwooldridge** [comment]
@greenhost87 If you can provide me a stack trace dump I would appreciate it. It is highly unlikely that loom threads would deadlock in HikariCP. If anything, the deadlock would be due to the driver, not HikariCP.
